File: services/deployment-service/src/deployment_service/infrastructure/messaging/publisher.py
from uuid import UUID

from faststream.kafka import KafkaBroker


class MessagePublisher:
    def __init__(self, broker: KafkaBroker):
        self._broker = broker
        self._logs_publisher = broker.publisher('service-logs')

    async def publish_deployment_created(
        self,
        deployment_id: UUID,
        project_id: UUID,
        config_id: UUID,
        version: str,
    ) -> None:
        await self._broker.publish(
            {
                'deployment_id': str(deployment_id),
                'project_id': str(project_id),
                'config_id': str(config_id),
                'version': version,
            },
            'deployment.created',
        )
        
        from datetime import datetime, UTC
        await self._logs_publisher.publish({
            "service": "deployment-service",
            "level": "INFO",
            "message": f"Deployment created: {deployment_id} for project {project_id}",
            "deployment_id": str(deployment_id),
            "project_id": str(project_id),
            "version": version,
            "action": "deployment.created",
            "timestamp": datetime.now(UTC).isoformat(),
            "environment": "development"
        })

    async def publish_deployment_building(
        self,
        deployment_id: UUID,
        project_id: UUID,
    ) -> None:
        await self._broker.publish(
            {
                'deployment_id': str(deployment_id),
                'project_id': str(project_id),
            },
            'deployment.building',
        )

    async def publish_deployment_deploying(
        self,
        deployment_id: UUID,
        project_id: UUID,
    ) -> None:
        await self._broker.publish(
            {
                'deployment_id': str(deployment_id),
                'project_id': str(project_id),
            },
            'deployment.deploying',
        )

    async def publish_deployment_running(
        self,
        deployment_id: UUID,
        project_id: UUID,
        image_url: str,
        deployment_url: str | None = None,
    ) -> None:
        await self._broker.publish(
            {
                'deployment_id': str(deployment_id),
                'project_id': str(project_id),
                'image_url': image_url,
                'deployment_url': deployment_url,
            },
            'deployment.running',
        )

        from datetime import datetime, UTC
        await self._logs_publisher.publish({
            "service": "deployment-service",
            "level": "INFO",
            "message": f"Deployment successful: {deployment_id}",
            "deployment_id": str(deployment_id),
            "project_id": str(project_id),
            "deployment_url": deployment_url,
            "action": "deployment.success",
            "timestamp": datetime.now(UTC).isoformat(),
            "environment": "development"
        })

    async def publish_deployment_failed(
        self,
        deployment_id: UUID,
        project_id: UUID,
        error_message: str,
    ) -> None:
        await self._broker.publish(
            {
                'deployment_id': str(deployment_id),
                'project_id': str(project_id),
                'error_message': error_message,
            },
            'deployment.failed',
        )

        from datetime import datetime, UTC
        await self._logs_publisher.publish({
            "service": "deployment-service",
            "level": "ERROR",
            "message": f"Deployment failed: {deployment_id} - {error_message}",
            "deployment_id": str(deployment_id),
            "project_id": str(project_id),
            "error_message": error_message,
            "action": "deployment.failed",
            "timestamp": datetime.now(UTC).isoformat(),
            "environment": "development"
        })

File: services/deployment-service/internal/infra/kafka/publisher.go
package kafka

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/twmb/franz-go/pkg/kgo"
)

type Publisher struct {
	client *kgo.Client
}

func NewPublisher(client *kgo.Client) *Publisher {
	return &Publisher{client: client}
}

func (p *Publisher) PublishBuilding(ctx context.Context, deploymentID, projectID uuid.UUID) error {
	return p.publish(ctx, "deployment.building", map[string]any{
		"deployment_id": deploymentID.String(),
		"project_id":    projectID.String(),
	})
}

func (p *Publisher) PublishDeploying(ctx context.Context, deploymentID, projectID uuid.UUID) error {
	return p.publish(ctx, "deployment.deploying", map[string]any{
		"deployment_id": deploymentID.String(),
		"project_id":    projectID.String(),
	})
}

func (p *Publisher) PublishRunning(ctx context.Context, deploymentID, projectID uuid.UUID, imageURL, deploymentURL string) error {
	return p.publish(ctx, "deployment.running", map[string]any{
		"deployment_id":  deploymentID.String(),
		"project_id":     projectID.String(),
		"image_url":      imageURL,
		"deployment_url": deploymentURL,
	})
}

func (p *Publisher) PublishFailed(ctx context.Context, deploymentID, projectID uuid.UUID, reason string) error {
	return p.publish(ctx, "deployment.failed", map[string]any{
		"deployment_id": deploymentID.String(),
		"project_id":    projectID.String(),
		"error_message": reason,
	})
}

func (p *Publisher) publish(ctx context.Context, topic string, payload map[string]any) error {
	payload["timestamp"] = time.Now().UTC().Format(time.RFC3339)

	data, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("marshal event %q: %w", topic, err)
	}

	results := p.client.ProduceSync(ctx, &kgo.Record{
		Topic: topic,
		Value: data,
	})
	if err := results.FirstErr(); err != nil {
		return fmt.Errorf("publish %q: %w", topic, err)
	}
	return nil
}
